### PR TITLE
protocol/bc: fix tx sighash and witness hash bugs

### PR DESCRIPTION
**Pull request by bobg**
Splits `OutputCommitment.writeTo` into explicit `writeExtensibleString` and `writeContents` methods for clarity.

Closes https://github.com/chain/chain/issues/330 - sighash was wrongly including the output commitment's extensible-string prefix length.
Closes https://github.com/chain/chain/issues/335 - witness hash was wrongly omitting hash of (empty) witness.

### CODE REVIEW

**Comment by bobg** [comment]
PTAL

**Comment by bobg** [comment]
(Note, this splits out the bugfixing part of #349.)

**Comment by tessr** [comment]
commit message nit (com-nit?): 

maybe add something like "remove wrapping extensible string from the output commitment hash and add transaction common witness fields in the transaction witness hash" to this commit message body so that a reader doesn't have to dive into those issues

File: protocol/bc/transaction.go [hunk 1]
```diff
@@ -232,6 +239,23 @@ func (tx *Tx) WitnessHash() (hash Hash, err error) {
 	return hash, nil
 }
 
+func (tx *TxData) commonWitnessHash() (h Hash, err error) {
```
Comment (tessr) [inline protocol/bc/transaction.go]: maybe add a comment that this is writing an empty common witness.

Comment (bobg) [inline protocol/bc/transaction.go]: It is today, but it may not be in the future. Added the suggested comment to `writeCommonWitness` instead.

**Pull request merged**